server/src/main/java/org/elasticsearch/search/query/QueryPhase.java

```diff
@@ -303,6 +345,74 @@ static boolean execute(SearchContext searchContext,
         }
     }
 
+     private static Query tryRewriteNumericLongSort(SearchContext searchContext, IndexReader reader) throws IOException {
+        // check sort
+        Sort sort = searchContext.sort().sort;
+        if (Sort.RELEVANCE.equals(sort)) return null;
+        if (Sort.INDEXORDER.equals(sort)) return null;
+        if (sort.getSort().length > 1) return null; // we need only a single sort
+        SortField sortField = sort.getSort()[0];
+
+        // check if this is a field of type Long, that is indexed and has doc values
+        String fieldName = sortField.getField();
+        final MappedFieldType fieldType = searchContext.mapperService().fullName(fieldName);
+        if (fieldType == null) return  null;
+        if (fieldType.typeName() != "long") return null;
+        if (fieldType.indexOptions() == IndexOptions.NONE) return null;
```

Comment:
this is always going to be NONE on long fields, which are indexed with points, you should check the point dimension count instead

Reply:
@jpountz  Actually for long fields, for `fieldType.indexOptions()` I am getting `IndexOptions.DOCS_AND_FREQS_AND_POSITIONS`  and for `fieldType.pointDataDimensionCount() ` I am getting 0.  Should we modify first `NumberFieldType` to have IndexOptions.NONE  and `dataDimensionCount = 1`?

Reply:
+1 to set the dimension count on numeric and date fields.

Reply:
oh sorry I got confused between what the actual fieldtype would look like at the Lucene level and what the Elasticsearch fieldtype looks like. Changing the MappedFieldType would be nice but from what I remember it might not be straightforward due to the fact that we have a fair number of functions that assume that indexOptions != NONE means "indexed". So let's not try to do it as part of this PR.